src/index.ts

```diff
@@ -0,0 +1,244 @@
+  defaultOptions: IAxiosRetry.IAxiosRetryConfig | undefined
+) {
+  const currentState = getRequestOptions(config, defaultOptions || {});
+  currentState.retryCount = currentState.retryCount || 0;
+  currentState.lastRequestTime = currentState.lastRequestTime || Date.now();
+  config[namespace] = currentState;
+  return currentState as Required<IAxiosRetry.IAxiosRetryConfigExtended>;
+}
+
+function fixConfig(axiosInstance: AxiosInstance | AxiosStatic, config: AxiosRequestConfig) {
+  // @ts-ignore
+  if (axiosInstance.defaults.agent === config.agent) {
+    // @ts-ignore
+    delete config.agent;
+  }
+  if (axiosInstance.defaults.httpAgent === config.httpAgent) {
+    delete config.httpAgent;
+  }
+  if (axiosInstance.defaults.httpsAgent === config.httpsAgent) {
+    delete config.httpsAgent;
+  }
+}
+
+async function shouldRetry(
+  currentState: Required<IAxiosRetry.IAxiosRetryConfig> & IAxiosRetry.IAxiosRetryConfigExtended,
+  error: AxiosError
+) {
+  const { retries, retryCondition } = currentState;
+  const shouldRetryOrPromise = (currentState.retryCount || 0) < retries && retryCondition(error);
+
+  // This could be a promise
+  if (typeof shouldRetryOrPromise === 'object') {
+    try {
+      const shouldRetryPromiseResult = await shouldRetryOrPromise;
+      // keep return true unless shouldRetryPromiseResult return false for compatibility
+      return shouldRetryPromiseResult !== false;
+    } catch (_err) {
+      return false;
+    }
+  }
+  return shouldRetryOrPromise;
+}
+
+const axiosRetry: IAxiosRetry.AxiosRetry = (axiosInstance, defaultOptions) => {
+  const requestInterceptorId = axiosInstance.interceptors.request.use((config) => {
+    setCurrentState(config, defaultOptions);
+    return config;
+  });
+
+  const responseInterceptorId = axiosInstance.interceptors.response.use(null, async (error) => {
+    const { config } = error;
+    // If we have no information to retry the request
+    if (!config) {
+      return Promise.reject(error);
+    }
+    const currentState = setCurrentState(config, defaultOptions);
+    if (await shouldRetry(currentState, error)) {
+      currentState.retryCount += 1;
+      const { retryDelay, shouldResetTimeout, onRetry } = currentState;
+      const delay = retryDelay(currentState.retryCount, error);
+      // Axios fails merging this configuration to the default configuration because it has an issue
+      // with circular structures: https://github.com/mzabriskie/axios/issues/370
+      fixConfig(axiosInstance, config);
+      if (!shouldResetTimeout && config.timeout && currentState.lastRequestTime) {
+        const lastRequestDuration = Date.now() - currentState.lastRequestTime;
+        const timeout = config.timeout - lastRequestDuration - delay;
+        if (timeout <= 0) {
+          return Promise.reject(error);
+        }
+        config.timeout = timeout;
+      }
+      config.transformRequest = [(data) => data];
+      await onRetry(currentState.retryCount, error, config);
```

Comment:
Here typescript tells me that the `await` has no effect. So either it should be removed or the definition of `onRetry` changed to return `Promise<void> | void`.

Might be tricky to make a call here. It's possible that people who ignored types before have actually relied on this returning a promise... People who used type checking obviously didn't as that would cause an error.

Reply:
yeah it seems we forgot to adjust the type.
Strange thing is that I didn't see that warning in VSCode (but only when I hove the line), what am I missing?

Reply:
This is reported as a "hint" by typescript so it's not very apparent. 

Noticed it manually when my editor showed it this way:

![Screenshot 2023-11-19 at 14 53 51](https://github.com/softonic/axios-retry/assets/153197/f8188d34-021c-4280-82da-0a62629e238e)